#12390 Add support for timeouts for functions added via addCleanup

p12tic opened the pull request:
## Scope and purpose

Fixes #12390

User-defined test functionality is currently run via `TestCase._run` to support terminating the user code if test case timeout is exceeded. The only exception is functions added via `addCleanup`. This PR changes `TestCase` to run these functions via `_run` too thus supporting timeouts for them.

exarkun commented on src/twisted/trial/_asynctest.py:
```diff
@@ -58,6 +58,12 @@ def __init__(self, methodName="runTest"):
+        # patching to the reactor done by the user.
+        from twisted.internet import reactor
+
+        self.__reactor = reactor
```
Twisted doesn't use the double underscore prefix.  Private attributes are just normal private attributes.

p12tic replied:
I agree in general, but in this particular case I think it's worth considering double prefix. 

TestCase class is by the users, so any attribute we use may clash with with anything users sets on the instance. It is especially true for `_reactor`, because users who write tests with some kind of mock reactor are likely to store it into test case attributes with similar name. I've just looked into the project I maintain - Buildbot - and there's already cases where `_reactor` is set on the test case.

exarkun replied:
Just pick a different name, then

p12tic replied:
Renamed to `_twistedPrivateReactor`, let's hope the new name is not too verbose.

I think the rest of attributes should have similar names.

exarkun commented on src/twisted/trial/test/detests.py:
```diff
@@ -186,6 +186,24 @@ def test_timeoutZero(self):
+
+        self.addCleanup(cleanup)
+
+    test_addCleanupTimeout.timeout = 0.1  # type: ignore[attr-defined]
```
Ideally, this functionality would be tested against a test double clock so that it could complete nearly instantly (instead of always taking at least 100ms), be robust against inconvenient scheduler decisions maybe the OS, and wouldn't hang forever if the functionality isn't working correctly.

A lot of the existing trial tests are not written in the ideal style.  If you can easily figure out how to rewrite this test, it's worth doing so but if it's not obvious how to do it now then it would also be reasonable to re-visit as a follow-up (not blocking this fix on it).  The more trial maintenance you want to do, the more important it will be to figure out how to write tests in the preferred way, though.

p12tic committed:
Pass method to TestCase._run directly

This allows more flexibility for _run to operate on any kind of
function.

p12tic committed:
Do not presume that TestCase._run needs to accept a method

_run() can support any function now, so the parameter names should
reflect that.

p12tic committed:
Acquire and store real reactor before running TestCase tests

This ensures that any reactor patching done by the user does not affect
how TestCase runs. TestCase uses the reactor only to support timeouts,
so this should be the real reactor that tracks real time.

p12tic committed:
Run cleanups via TestCase._run() to support timeouts

Currently cleanup functions registered via addCleanup() are run
directly, without taking into account that they may take indeterminate
time to complete. Running the functions through _run allows to terminate
the test if the cleanups take too long.

p12tic commented:
please review

adiroiban commented on src/twisted/trial/test/test_deferred.py:
```diff
@@ -210,6 +210,19 @@ def test_timeoutZero(self) -> None:
         assert isinstance(result.errors[0][1], Failure)
         self._wasTimeout(result.errors[0][1])
 
+    def test_addCleanupPassDefault(self) -> None:
```
Can we add a docstring explaining why we have this test?
What is the test scenario here?

Same comment for `test_addCleanupTimeou`

p12tic replied:
Fixed, thanks.

adiroiban commented on src/twisted/trial/test/detests.py:
```diff
@@ -186,6 +186,24 @@ def test_timeoutZero(self):
     test_timeoutZero.timeout = 0  # type: ignore[attr-defined]
 
+    def test_addCleanupPassDefault(self):
+        # test default timeout for addCleanup functions
```
Can we have this comment as a docstring?

What does `default timeout` means ?

From `test_addCleanupPassDefault` I understand that "this is a test about cleanup with passing a default" 

Maybe something likes

"""
A cleanup can return a deferred.
The cleanup is successuful as long as the deferred
is resolve sooner than the default test case
timeout our 5 seconds.
"""

p12tic replied:
Fixed, thanks.

adiroiban commented on src/twisted/trial/test/detests.py:
```diff
@@ -186,6 +186,24 @@ def test_timeoutZero(self):
+        self.addCleanup(cleanup)
+
+    def test_addCleanupTimeout(self):
+        # test that functions added via addCleanup supports timeout
```
Can we have this as a docstring ?

What does "support timeout" mean ?

Maybe something like this

```suggestion
        """
        When the deferred returned by addCleanup is not
        resolved sooner than the test's timeout,
        the test is considered failed.
        
        For this test, we retrun from C{addCleanup} a
        deferred that will never receive a result.
```

p12tic replied:
Fixed.

adiroiban commented on src/twisted/trial/test/test_deferred.py:
```diff
@@ -210,6 +210,19 @@ def test_timeoutZero(self) -> None:
+        self.assertEqual(result.testsRun, 1)
+
+    def test_addCleanupTimeout(self) -> None:
+        result = self.runTest("test_addCleanupTimeout")
```
Maybe we can have it written like this

with a docstring and a test-double scheduler.

```suggestion
        """
        See L{twisted.trial.test.detests.TimeoutTests.test_addCleanupTimeout}
        """
        clock = task.Clock()
        self._twistedPrivateReactor = clock

        result = self.runTest("test_addCleanupTimeout")

        clock.advance(DEFAULT_TIMEOUT)
```

p12tic replied:
Unfortunately this is not possible because runTest() eventually calls into TestCase._runFixturesAndTest where it waits via TestCase._wait() (which runs reactor inside) until the test either fails or succeeds. It would require quite a lot of refactoring to support fake reactor in this test.

adiroiban commented on src/twisted/trial/_asynctest.py:
```diff
@@ -58,6 +58,14 @@ def __init__(self, methodName="runTest"):
+
+        # A unique name must be used in order not to clash with user code that
+        # sets their own attributes.
+        self._twistedPrivateReactor = reactor
```
As this instance is only used to schedule tasks and not do do any I/O operations, maybe this can be called a "scheduler"

```suggestion
        self._twistedPrivateScheduler = reactor
```

p12tic replied:
Fixed, thanks.

adiroiban commented on src/twisted/trial/newsfragments/12390.bugfix:
```diff
@@ -0,0 +1 @@
+Functions added via ``TestCase.addCleanup`` now are executed taking test case timeout into account.
```
A minor sugestion to get the release notes with the canonical name of the targeted API 

```suggestion
twisted.trial.unittest.TestCase.addCleanup will cause the test to fail if the returned deferred is not resolved before the test's timeout.
```

p12tic replied:
Fixed, thanks.

adiroiban commented on src/twisted/trial/test/test_deferred.py:
```diff
@@ -210,6 +210,19 @@ def test_timeoutZero(self) -> None:
+        self.assertEqual(result.testsRun, 1)
+        self.assertEqual(len(result.errors), 1)
+        assert isinstance(result.errors[0][1], Failure)
+        self._wasTimeout(result.errors[0][1])
```
I see that `_wasTimeout` only checks for 

```
self.assertEqual(error.check(defer.TimeoutError), defer.TimeoutError)
```

Can we also check that "cleanup function" is part of the error message, to make sure the timeout is not accidentally triggered by someting else ?

p12tic replied:
Fixed now.

adiroiban reviewed (changes requested):
Many thanks for the fix.

The changes looks very good.

I left a few minor comments regarding the new tests.

I know that the existing tests have no docstrings or description of why they exists.

I think that for the new tests that are written, we should follow the current best practices and add more info about why we wrote those tests and that are they trying to tests.

There are no tests for the test code itself, so if a test fails, a docstring can help to identify whether the test failure is due an error in the targeted code, or the test case is not correct.

Thanks again

p12tic committed:
Verify timeout exception messages in TestCase

p12tic commented:
@adiroiban I've addressed your comments.

p12tic commented:
This PR should be merged after https://github.com/twisted/twisted/pull/12393

adiroiban reviewed (approved):
Thanks for the update.

I think that this is a big improvement over the existing code and we can merge this.

Thanks again

p12tic commented:
The regression is somewhat expected, but should be ignored. Further PRs will optimize test runtime by more than 10%.

glyph commented:
I acknowledged the regressions since they all appear to be quite specific to trial, and I'm sure that `addCleanup` doesn't dominate test runtime. (Interesting abstract question, does this perceptibly increase our own full test suite runtime?)

p12tic commented:
> Interesting abstract question, does this perceptibly increase our own full test suite runtime?

No. Even the performance tests are affected just by few percent. They don't do anything except call `addCleanup()` so they're pretty much worst case.

p12tic commented:
Just noticed inaccuracy in previous comment:

> test runtime by more than 10%.

should be "performance test runtime".
